### ISSUE TITLE
scrypted docker is deleting or changing the host /run/dbus/system_bus_socket, killing host services that depend on dbus

### ISSUE THREAD

**Issue by douginoz** [koush/scrypted#936]
The scrypted docker compose file maps the host's /var/run/dbus directory to it's own.  Scrypted's docker initialisation in `install/docker/fs/etc/s6-overlay/scripts/setup.sh` (line 12) then removes any existing /var/run/dbus/system_bus_socket file it finds and creates it's own, and line 15  changes permission at the same time.  Any service on the host using dbus then hangs/aborts.  

Scrypted should never be deleting or chowning the host /run/dbus/system_bus_socket, should it? 

To reproduce this:
-  stop/remove the scrypted container on a ubuntu system
-  Install bluez, and confirm that the bluetooth service is running
- run 'bluetoothctl'.  Bluetoothctl uses dbus to communicate with the bluetooth service.  It should run fine. ('quit' to exit).
- Now start up the scrypted container then try running bluetoothctl again on the host.  Scrypted deletes the current host /run/dbus/system_bus_socket file.  Bluetoothctl will fail to start.
 
Stopping and restarting the bluetooth service also fails at this point because the exiting dbus connection it had been using has been removed by scrypted and replaced with one it can't use.  Possibly because the permissions have changed on the folder or file.

Even if scrypted is stopped at this point, the bluetooth service can't be restarted because scrypted has removed the original /run/dbus/system_bus_socket on the host that it had access to.

To get it working again, the dbus.service and dbus.socket service must be stopped then restarted, which re-creates a new /run/dbus/system_bus_socket file.  However, as soon as scrypted is started again, it deletes the host file and creates it's own.

**Comment by wlharr04** [comment]
I'm experiencing what I'm assuming to be the exact same issue, but wasn't sure of the specific cause. As soon as I read your post, everything just clicked. All I could figure out before then is that while the scrypted container was running, somehow it completely destroyed my ability to use avahi and start new containers in Docker because of dbus issues.

**Comment by koush** [comment]
will fix

**Comment by koush** [comment]
I'm not seeing where /var/run/dbus is mapped?

**Comment by koush** [comment]
Oh, you're modifying the docker compose to enable the mount. I've never actually tested that avahi stuff. It was user contributed.

**Comment by douginoz** [comment]
That’s definitely a symptom. if my Scrypted container started before all the others were running, subsequent ones fail to start. Portainer returned a 500 error which means little. Since my other containers are all years old and stable I suspected the Scrypted one. And discovered that if I run it last, I can get every container running. But since there’s no way to prioritize container start sequence I needed to isolate what’s going on. The Bluetooth failure messages led me to the dbus service and that led me back to the line in the Scrypted docker compose yaml. That got me looking at the Scrypted source and I found the strange code block that I mentioned before, that appears to first delete any existing /run/dbus file then change the parent directory owner. But since this is actually linked to the host files in the docker compose, it’s touching the system wide messaging file and not just one used within the container. That seems like an error but I’m not really sure what Scrypted is attempting to do to know if this is a valid action. I suspect it’s not. And if it’s not, then I suspect the fix will be for Scrypted to use it’s own application-specific dbus file or open any existing file with appropriate permissions so that it allows other processes to continue to use it. The error most likely occurred because the actual code was written without docker in mind, and later on when the docker compose file was created, the dbus folder was mapped to the host without remembering that Scrypted deletes it. There’s been quite a lot of discussion in discord relating to avahi and dbus so perhaps the code was inserted to fix another problem and it’s only now that downstream issues are being noticed. On Jul 8, 2023, at 22:37, wlharr04 ***@***.***> wrote:﻿
I'm experiencing what I'm assuming to be the exact same issue, but wasn't sure of the specific cause. As soon as I read your post, everything just clicked. All I could figure out before then is that while the scrypted container was running, somehow it completely destroyed my ability to use avahi and start new containers in Docker because of dbus issues.

**Comment by douginoz** [comment]
Yes because the docker compose says to enable the next two lines in order to use avahi. I know that avahi is used in some way to allow devices (especially Apple) to discover each other so I assumed that it’s required in order for Scrypted to find my cameras or for other systems to discover Scrypted. Since I’m using Scrypted for use with HomeKit, which uses avahi, it seems likely that this is needed. There needs to be something detailing when it’s required. There’s another line in the docker compose that I have no idea what it’s used for - the /dev/dri line - so i don’t know what the impact of leaving it in or taking it out has on getting things up and running. Without some sort of information on these things it makes it very difficult to figure out why something isn’t working or whether there’s anything I can change to get things working much better. On Jul 8, 2023, at 23:25, Koushik Dutta ***@***.***> wrote:﻿
Oh, you're modifying the docker compose to enable the mount. I've never actually tested that avahi stuff. It was user contributed, which is why its not enabled by default.

**Comment by qosmio** [comment]
> Oh, you're modifying the docker compose to enable the mount. I've never actually tested that avahi stuff. It was user contributed, which is why its not enabled by default.

Since the example mentions setting `network_mode: host`, we shouldn't need to mount `/run/dbus` inside the container. After removing the volume mount, I was able to discover my Reolink WiFi Doorbell in Scrypted as well as in the Home app after enabling the extension for it.

**Comment by wlharr04** [comment]
> > Oh, you're modifying the docker compose to enable the mount. I've never actually tested that avahi stuff. It was user contributed, which is why its not enabled by default.
> 
> Since the example mentions setting `network_mode: host`, we shouldn't need to mount `/run/dbus` inside the container. After removing the volume mount, I was able to discover my Reolink WiFi Doorbell in Scrypted as well as in the Home app after enabling the extension for it.

That makes perfect sense to me! I don't know why I never realized there would be any conflict with that. I just assumed code gremlins!

**Comment by douginoz** [comment]
> Hey @douginoz
> 
> wanna give the latest change a try?

I'm not sure!  What's the latest change?  Do you mean what you did over on #940?  

I see that the latest version of scrypted was from yesterday - does that include those changes and is that the one I should be trying?

But before that, looking at #940 (I should probably comment over there), and keeping in mind that I'm not very knowledgable on much of it, it seems that you still want to delete the host's dbus.  You don't, as you've said, but you still want to (as it relates to avahi or something).  

That seems weird.  Shouldn't an app either use it's own (host) dbus file, or share any existing one?  I can't figure out why scrypted needs to create and have exclusive use of /run/dbus/system_bus_socket.  It's kind of like an app insisting that it needs exclusive access to the system boot drive ("\").

**Comment by douginoz** [comment]
How do I test this out?  I installed scrypted via the supported docker compose.  Is there a different image to try?  
I'm afraid I'm not a dev so I'm not set up with a test system.

**Comment by koalaeagle** [comment]
Damn, this one hit me hard. I enabled this since my cameras constantly give offline and then online notifications, and I saw that avahi was mentioned as the best experience for Linux.

Enabling this mount didn't seem to cause any issues at first, and I happened to migrate to a different distro ~1 week after. Couldn't figure out why I had vague dbus errors causing many things to fail, which I assumed was related to the new distro. Didn't realise I was causing a self-ddos on my system with this one anytime the scrypted container came up. For me, it seemed to cause permanent damage to dbus and even with scrypted removed and the system rebooted I had dbus issues and had to re-install.

**Comment by koush** [comment]
Per the homebridge docs from where this pull request was inspired, you should *not* be passing in the entire /var/run into the container.
https://github.com/homebridge/homebridge/wiki/mDNS-Options#avahi-linux-and-docker

Only the directories:

```yaml
    security_opt:
      - apparmor:unconfined
    volumes:
      - /var/run/dbus:/var/run/dbus
      - /var/run/avahi-daemon/socket:/var/run/avahi-daemon/socket
      - ...
```

The avahi startup script will delete the running pids in the container. If the docker compose is passing in the entire host `/var/run` then the host pids are what are removed:

```sh
# delete existing pids if they exist
[ -e /var/run/dbus.pid ] && rm -f /var/run/dbus.pid
[ -e /var/run/dbus/pid ] && rm -f /var/run/dbus/pid 
[ -e /run/dbus/pid ] && rm -f /run/dbus/pid
[ -e /var/run/avahi-daemon/pid ] && rm -f /var/run/avahi-daemon/pid
[ -e /var/run/dbus/system_bus_socket ] && rm -f /var/run/dbus/system_bus_socket
```

I'll add documentation sample to the docker-compose.yml.

**Comment by ethanspitz** [comment]
This is still very much a problem for me. I just spent the last few hours trying to figure out dbus issues not realizing that scrypted was the root cause. I had ` - /var/run/dbus:/var/run/dbus` in my docker compose which resulted in truenas having tons and tons of issues. Once I removed the line from docker-compose, everything was all good. I don't think I needed the functionality anyway for scrypted, but I don't think the example should include it by default.
